### DISCUSSION TITLE
Populate function does not correctly derive property type

### DISCUSSION THREAD

**Discussion by AlexRies** [typegoose/typegoose#488] [Q&A] [answered]
Hi everyone,

I encountered the following problem with using the mongoose populate function.
In the following example the populate function is used to populate the `User.car` property.
Please fix or provide guidance how to derive the `User.car` property type correctly. Thank you!

Observed behavior:
- The TypeScript type of `foundUser?.car` is derived as `Car | ObjectId`
![image](https://user-images.githubusercontent.com/77303034/107954558-e3102700-6f9c-11eb-946b-a02e9e80f7c7.png)

Expected behavior:
- The TypeScript type of `foundUser?.car` should be derived as `Car`

Example:
```ts
import {
  prop,
  getModelForClass,
  Ref,
  modelOptions,
} from "@typegoose/typegoose";
import mongoose from "mongoose";

@modelOptions({ schemaOptions: { collection: "cars" } })
class Car {
  @prop()
  public model?: string;
}

@modelOptions({ schemaOptions: { collection: "user" } })
class User {
  @prop({ ref: () => Car, required: true })
  public car?: Ref<Car>;
}

(async () => {
  await mongoose.connect("mongodb://localhost:27017/", {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    dbName: "test",
  });

  const CarModel = getModelForClass(Car);
  const UserModel = getModelForClass(User);

  const newCar: Car = {
    model: "Porsche",
  };

  const createdCar = await CarModel.create(newCar);
  const newUser: User = {
    car: createdCar.id,
  };
  const createdUser = await UserModel.create(newUser);

  const foundUser = await UserModel.findById(createdUser.id)
    .populate("car")
    .exec();

  const model = foundUser?.car?.model;
  console.log(model);
})();

```

**Comment by hasezoey** [comment, marked as answer]
you probably search for [`isDocument` typeguard](https://typegoose.github.io/typegoose/docs/api/functions/typeguards/is-document)

**Reply by AlexRies** [reply]
That is a workaround, but it should be possible to return the correct type for `User.car`, depending whether populate was called or not on the `UserModel`. Is there a plan to support that?

**Reply by hasezoey** [reply]
this is not an workaround, before calling this, it can be either the wanted document or an id, and populate cannot change the type on its own (at least not without changing the full type, which typegoose will not do - typegoose will not modify an mongoose functions)
